napari/components/viewer_model.py

```diff
@@ -355,14 +355,37 @@ def rounded_division(min_val, max_val, precision):
         return int(((min_val + max_val) / 2) / precision) * precision
 
     def _on_layers_change(self):
+        new_point = None
         if len(self.layers) == 0:
             self.dims.ndim = 2
             self.dims.reset()
         else:
+            prev_point = self.dims.point
             ranges = self.layers._ranges
             ndim = len(ranges)
             self.dims.ndim = ndim
             self.dims.set_range(range(ndim), ranges)
+            if len(prev_point) == len(self.dims.point):
+                is_in_range = all(
+                    [
+                        range_vals[0] <= point_vals < range_vals[1]
+                        for point_vals, range_vals in zip(prev_point, ranges)
+                    ]
+                )
+            else:
+                is_in_range = False
+
+            if not is_in_range:
+                new_point = [
+                    self.rounded_division(*_range) for _range in ranges
+                ]
+
+            elif self.dims.point != prev_point:
+                new_point = (
+                    prev_point[:ndim]
+                    if ndim < len(prev_point)
+                    else np.append(prev_point, (0,) * (ndim - len(prev_point)))
+                )
```

Comment:
This logic is fairly hard to read and reason about - sometimes we reset all the point's coordinates to be in the middle of the ranges and sometimes we pad with zeros. The existing logic around adding a new layer and its effect on `dims` is already complex, so I'd prefer something simpler here.

What about just keeping the point in the leading dimensions and padding with mid-points for new dimensions? For example:

```python
if ndim < len(prev_point):
    new_point = prev_point[:ndim]
else:
    mid_points = [self.rounded_division(*_range) for range in ranges[len(prev_point):]]
    new_point = np.append(prev_point, mid_points)
```

That removes some of the current `is_in_range` logic (i.e. when `ndim` hasn't changed) and will use the clamping behavior defined in `set_current_step` (which is always called by `set_point`) when the new point is no longer in the range of one of the dimensions. I think that extra logic might be handling the code you deleted that handles the special case of going from 0 to 1 layers - but maybe we can just bring that special case back instead?

Reply:
I agree that this is much clearer. However, I left some of the range logic inside, since the logic called by set_point only clamps points if they are below the lower range, but not if they are above the upper range.
I added a new test related to the Points layer, which addresses this issue specifically. :)

I could also look into set_current_step to add the upper-bound logic there? I did not want to touch too many functions though ^^

Reply:
I just realized that new dimensions are prepended rather than appended, therefore I changed the orders here :)